File: cineva/packages/audio_engine/lib/src/dsp/loudness.dart
/// Loudness ITU-R BS.1770-4 — miroir de `native/cineva_dsp/src/loudness.c`.
library;

import 'dart:math' as math;

import 'biquad.dart' as dsp;
import 'biquad.dart' show Biquad;

const int _loudnessHops = 600;
const int _loudnessSilenceBlocks = 10;
const double _loudnessOffset = -0.691;

class LoudnessProcessor {
  LoudnessProcessor(this.sampleRate) {
    hopLen = (sampleRate * 0.1).toInt();
    if (hopLen < 1) hopLen = 1;
    for (int i = 0; i < dsp.maxChannels; i++) {
      k1.add(Biquad());
      k2.add(Biquad());
      k1[i].init();
      k2[i].init();
      k1[i].set(2, 1681.97, 3.9998, 0.7071752369556, sampleRate); // high shelf
      k2[i].set(4, 38.13, 0, 0.5003270372833, sampleRate); // high pass RLB
      k1[i].snap();
      k2[i].snap();
      hopMs.add(0);
    }
    hopZ = List<double>.filled(_loudnessHops, 0);
    set(true, -16, 8, 8, 1.5);
  }

  final double sampleRate;
  bool enable = false;
  double targetLufs = -16;
  double maxGainDb = 8;
  double maxAttDb = 8;
  double adaptRateDbS = 1.5;

  late int hopLen;
  int hopPos = 0;
  final List<double> hopMs = <double>[];
  final List<Biquad> k1 = <Biquad>[];
  final List<Biquad> k2 = <Biquad>[];
  late List<double> hopZ;
  int hopCount = 0;
  int hopHead = 0;
  int silentRun = 0;

  double integratedLufs = -70;
  double gainDb = 0;
  double targetGainDb = 0;

  void set(bool enable, double targetLufs, double maxGainDb, double maxAttDb,
      double adaptRateDbS) {
    this.enable = enable;
    this.targetLufs = dsp.clampD(dsp.sanitizeD(targetLufs), -36, -8);
    this.maxGainDb = dsp.clampD(dsp.sanitizeD(maxGainDb), 0, 12);
    maxAttDb = dsp.clampD(dsp.sanitizeD(maxAttDb), 0, 12);
    this.maxAttDb = maxAttDb;
    this.adaptRateDbS = dsp.clampD(dsp.sanitizeD(adaptRateDbS), 0.1, 6);
  }

  void reset() {
    hopPos = 0;
    for (int i = 0; i < dsp.maxChannels; i++) {
      hopMs[i] = 0;
    }
    hopZ = List<double>.filled(_loudnessHops, 0);
    hopCount = 0;
    hopHead = 0;
    silentRun = 0;
    integratedLufs = -70;
    gainDb = 0;
    targetGainDb = 0;
    for (int i = 0; i < dsp.maxChannels; i++) {
      k1[i].resetState();
      k2[i].resetState();
    }
  }

  void _recompute() {
    final int n = hopCount;
    if (n <= 0) {
      integratedLufs = -70;
      return;
    }
    double sum = 0;
    int count = 0;
    for (int i = 0; i < n; i++) {
      final double z = hopZ[i];
      if (z <= 0) continue;
      final double lb = _loudnessOffset + 10 * _log10(z);
      if (lb > -70) {
        sum += z;
        count++;
      }
    }
    if (count == 0 || sum <= 0) {
      integratedLufs = -70;
      return;
    }
    final double meanZ = sum / count;
    final double relThreshold = _loudnessOffset + 10 * _log10(meanZ) - 10;
    double sum2 = 0;
    int count2 = 0;
    for (int i = 0; i < n; i++) {
      final double z = hopZ[i];
      if (z <= 0) continue;
      final double lb = _loudnessOffset + 10 * _log10(z);
      if (lb > -70 && lb > relThreshold) {
        sum2 += z;
        count2++;
      }
    }
    if (count2 == 0 || sum2 <= 0) {
      integratedLufs = -70;
      return;
    }
    integratedLufs = _loudnessOffset + 10 * _log10(sum2 / count2);
  }

  static double _log10(double x) => math.log(x) / math.ln10;

  void process(List<Float32List> in_, List<Float32List> out, int channels,
      int frames, List<double> weights) {
    if (channels < 1 || channels > dsp.maxChannels) return;
    final double ramp = 1 - dsp.onepoleCoef(sampleRate, 0.02);
    final double gainStepDb = adaptRateDbS / sampleRate;

    for (int ch = 0; ch < channels; ch++) {
      if (!identical(out[ch], in_[ch])) {
        for (int i = 0; i < frames; i++) {
          out[ch][i] = in_[ch][i];
        }
      }
    }

    if (!enable) {
      gainDb = 0;
      targetGainDb = 0;
      return;
    }

    for (int i = 0; i < frames; i++) {
      for (int ch = 0; ch < channels; ch++) {
        final double x = dsp.sanitizeD(in_[ch][i]);
        double y = k1[ch].tick(x, ramp);
        y = k2[ch].tick(y, ramp);
        hopMs[ch] += y * y;
      }

      hopPos++;
      if (hopPos >= hopLen) {
        hopPos = 0;
        double z = 0;
        for (int ch = 0; ch < channels; ch++) {
          final double ms = hopMs[ch] / hopLen;
          z += weights[ch] * ms;
          hopMs[ch] = 0;
        }
        hopZ[hopHead] = z;
        hopHead = (hopHead + 1) % _loudnessHops;
        if (hopCount < _loudnessHops) hopCount++;

        if (z < 1e-10) {
          silentRun++;
        } else {
          silentRun = 0;
        }
        _recompute();

        if (silentRun >= _loudnessSilenceBlocks || integratedLufs <= -69.9) {
          targetGainDb = gainDb; // gel pendant le silence
        } else {
          final double want = targetLufs - integratedLufs;
          targetGainDb = dsp.clampD(want, -maxAttDb, maxGainDb);
        }
      }

      if (gainDb < targetGainDb) {
        gainDb += gainStepDb;
        if (gainDb > targetGainDb) gainDb = targetGainDb;
      } else if (gainDb > targetGainDb) {
        gainDb -= gainStepDb;
        if (gainDb < targetGainDb) gainDb = targetGainDb;
      }

      final double g = dsp.dbToLin(gainDb);
      for (int ch = 0; ch < channels; ch++) {
        out[ch][i] = dsp.sanitizeD(out[ch][i]) * g;
      }
    }
  }
}

File: cineva/packages/audio_engine/native/cineva_dsp/src/loudness.c
#include "loudness.h"

#include <math.h>
#include <string.h>

#include "dsp_util.h"

#define CV_LOUDNESS_OFFSET (-0.691)

void cv_loudness_init(cv_loudness *ln, double sample_rate) {
  memset(ln, 0, sizeof(*ln));
  ln->sample_rate = cv_clamp_d(sample_rate, 8000.0, 192000.0);
  ln->hop_len = (int)(ln->sample_rate * 0.1);
  if (ln->hop_len < 1) ln->hop_len = 1;
  ln->target_lufs = -16.0;
  ln->max_gain_db = 8.0;
  ln->max_att_db = 8.0;
  ln->adapt_rate_db_s = 1.5;
  ln->gain_db = 0.0;
  ln->target_gain_db = 0.0;
  ln->integrated_lufs = -70.0;
  for (int i = 0; i < CINEVA_DSP_MAX_CHANNELS; i++) {
    cv_biquad_init(&ln->k1[i]);
    cv_biquad_init(&ln->k2[i]);
    /* Pondération K : shelf +4 dB @ 1681.97 Hz (Q 0.7071) puis HP 38.13 Hz (Q 0.5). */
    cv_biquad_set(&ln->k1[i], CV_BQ_HIGHSHELF, 1681.97, 3.9998, 0.7071752369556, ln->sample_rate);
    cv_biquad_set(&ln->k2[i], CV_BQ_HIGHPASS, 38.13, 0.0, 0.5003270372833, ln->sample_rate);
    cv_biquad_snap(&ln->k1[i]);
    cv_biquad_snap(&ln->k2[i]);
  }
}

void cv_loudness_set(cv_loudness *ln, int enable, double target_lufs,
                     double max_gain_db, double max_att_db,
                     double adapt_rate_db_s) {
  ln->enable = enable ? 1 : 0;
  ln->target_lufs = cv_clamp_d(cv_sanitize_d(target_lufs), -36.0, -8.0);
  ln->max_gain_db = cv_clamp_d(cv_sanitize_d(max_gain_db), 0.0, 12.0);
  ln->max_att_db = cv_clamp_d(cv_sanitize_d(max_att_db), 0.0, 12.0);
  ln->adapt_rate_db_s = cv_clamp_d(cv_sanitize_d(adapt_rate_db_s), 0.1, 6.0);
}

void cv_loudness_reset(cv_loudness *ln) {
  ln->hop_pos = 0;
  for (int i = 0; i < CINEVA_DSP_MAX_CHANNELS; i++) ln->hop_ms[i] = 0.0;
  memset(ln->hop_z, 0, sizeof(ln->hop_z));
  ln->hop_count = 0;
  ln->hop_head = 0;
  ln->silent_run = 0;
  ln->integrated_lufs = -70.0;
  ln->gain_db = 0.0;
  ln->target_gain_db = 0.0;
  for (int i = 0; i < CINEVA_DSP_MAX_CHANNELS; i++) {
    cv_biquad_reset_state(&ln->k1[i]);
    cv_biquad_reset_state(&ln->k2[i]);
  }
}

/* Recalcule le loudness intégré gate (double gate) sur l'historique. */
static void cv_loudness_recompute(cv_loudness *ln) {
  int n = ln->hop_count;
  if (n <= 0) {
    ln->integrated_lufs = -70.0;
    return;
  }
  /* Passe 1 : gate absolu à -70 LUFS. */
  double sum = 0.0;
  int count = 0;
  for (int i = 0; i < n; i++) {
    double z = ln->hop_z[i];
    if (z <= 0.0) continue;
    double lb = CV_LOUDNESS_OFFSET + 10.0 * log10(z);
    if (lb > -70.0) {
      sum += z;
      count++;
    }
  }
  if (count == 0 || sum <= 0.0) {
    ln->integrated_lufs = -70.0;
    return;
  }
  double mean_z = sum / (double)count;
  double rel_threshold = CV_LOUDNESS_OFFSET + 10.0 * log10(mean_z) - 10.0;
  /* Passe 2 : gate relatif à -10 LU. */
  double sum2 = 0.0;
  int count2 = 0;
  for (int i = 0; i < n; i++) {
    double z = ln->hop_z[i];
    if (z <= 0.0) continue;
    double lb = CV_LOUDNESS_OFFSET + 10.0 * log10(z);
    if (lb > -70.0 && lb > rel_threshold) {
      sum2 += z;
      count2++;
    }
  }
  if (count2 == 0 || sum2 <= 0.0) {
    ln->integrated_lufs = -70.0;
    return;
  }
  ln->integrated_lufs = CV_LOUDNESS_OFFSET + 10.0 * log10(sum2 / (double)count2);
}

void cv_loudness_process(cv_loudness *ln, const float *const *in,
                         float *const *out, int channels, int frames,
                         const double *weights) {
  if (channels < 1 || channels > CINEVA_DSP_MAX_CHANNELS) return;
  double ramp = 1.0 - cv_onepole_coef(ln->sample_rate, 0.02);
  double gain_step_db = ln->adapt_rate_db_s / ln->sample_rate;

  for (int ch = 0; ch < channels; ch++) {
    if (out[ch] != in[ch]) memcpy(out[ch], in[ch], (size_t)frames * sizeof(float));
  }

  if (!ln->enable) {
    ln->gain_db = 0.0;
    ln->target_gain_db = 0.0;
    return;
  }

  for (int i = 0; i < frames; i++) {
    for (int ch = 0; ch < channels; ch++) {
      double x = cv_sanitize_f(in[ch][i]);
      double y = cv_biquad_tick(&ln->k1[ch], x, ramp);
      y = cv_biquad_tick(&ln->k2[ch], y, ramp);
      ln->hop_ms[ch] += y * y;
    }

    ln->hop_pos++;
    if (ln->hop_pos >= ln->hop_len) {
      ln->hop_pos = 0;
      double z = 0.0;
      for (int ch = 0; ch < channels; ch++) {
        double ms = ln->hop_ms[ch] / (double)ln->hop_len;
        z += (weights ? weights[ch] : 1.0) * ms;
        ln->hop_ms[ch] = 0.0;
      }
      ln->hop_z[ln->hop_head] = z;
      ln->hop_head = (ln->hop_head + 1) % CV_LOUDNESS_HOPS;
      if (ln->hop_count < CV_LOUDNESS_HOPS) ln->hop_count++;

      if (z < 1e-10) {
        ln->silent_run++;
      } else {
        ln->silent_run = 0;
      }
      cv_loudness_recompute(ln);

      if (ln->silent_run >= CV_LOUDNESS_SILENCE_BLOCKS || ln->integrated_lufs <= -69.9) {
        ln->target_gain_db = ln->gain_db; /* gel pendant le silence */
      } else {
        double want = ln->target_lufs - ln->integrated_lufs;
        ln->target_gain_db = cv_clamp_d(want, -ln->max_att_db, ln->max_gain_db);
      }
    }

    /* Gain adaptatif borné en vitesse (pas de pompage). */
    if (ln->gain_db < ln->target_gain_db) {
      ln->gain_db += gain_step_db;
      if (ln->gain_db > ln->target_gain_db) ln->gain_db = ln->target_gain_db;
    } else if (ln->gain_db > ln->target_gain_db) {
      ln->gain_db -= gain_step_db;
      if (ln->gain_db < ln->target_gain_db) ln->gain_db = ln->target_gain_db;
    }

    double g = cv_db_to_lin(ln->gain_db);
    for (int ch = 0; ch < channels; ch++) {
      out[ch][i] = (float)((double)cv_sanitize_f(out[ch][i]) * g);
    }
  }
}

double cv_loudness_integrated(const cv_loudness *ln) { return ln->integrated_lufs; }

double cv_loudness_gain_db(const cv_loudness *ln) { return ln->gain_db; }
